src/Build/BackEnd/Components/ResourceManager/ResourceManagerService.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Build.Shared;
+using System.Threading;
+
+#nullable enable
+
+namespace Microsoft.Build.BackEnd.Components.ResourceManager
+{
+    class ResourceManagerService : IBuildComponent
+    {
+        Semaphore? s = null;
+
+#if DEBUG
+        public int TotalNumberHeld = -1;
+        public string? SemaphoreName;
+#endif
+
+        internal static IBuildComponent CreateComponent(BuildComponentType type)
+        {
+            ErrorUtilities.VerifyThrow(type == BuildComponentType.TaskResourceManager, "Cannot create components of type {0}", type);
+
+            return new ResourceManagerService();
+        }
+
+        public void InitializeComponent(IBuildComponentHost host)
+        {
+            string semaphoreName = host.BuildParameters.ResourceManagerSemaphoreName;
+
+            int resourceCount = host.BuildParameters.MaxNodeCount; // TODO: tweakability
+
+#if DEBUG
+            TotalNumberHeld = 0;
+            SemaphoreName = semaphoreName;
+#endif
+
+            if (NativeMethodsShared.IsWindows)
```

Comment:
Since the current plan is to ignore non-windows, why initialize s at all? Also, can you move this to the top of this function to avoid doing unnecessary work? Note that giving it a value, as you do just below, means it will be treated as "non-null" as far as requesting/releasing cores --> the nullable changes you made won't have any effect.